### PR TITLE
[WIP] Rule update: Heading is descriptive (b49b2e)

### PR DESCRIPTION
**Pull request by annethyme**
@rvantonisse wrote in https://github.com/auto-wcag/auto-wcag/issues/424: 

> During several audits with this rule, “Heading is Descriptive”, we encountered some situations in which we were forced by the rule to fail the SC.
> 
> The stituations were:
> 
> - Empty heading with a section of the content.
> - Heading without section of the content
> - Heading that describes only a part of the section of the content.
> 
> We believe these situations should not fail the WCAG SC 2.4.6 and therefore the rule needs some improvements.

Closes issue: https://github.com/auto-wcag/auto-wcag/issues/424

Also updating the rule to:
- in the Applicability use the replacements terms for the "non-empty" definition introduced in this pull request: https://github.com/auto-wcag/auto-wcag/pull/430/files
- in the Expectation evaluate _both_ the accessible name and text nodes of the heading, if the two are different. This was not clear from the rule beforehand.

# How to Review And Approve
- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines. 
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.

### CODE REVIEW

File: _rules/SC2-4-6-descriptive-headings.md [hunk 1]
```diff
@@ -22,13 +23,13 @@ authors:
 
 ### Applicability
 
-This rule applies to any element with the [semantic role](#semantic-role) of heading that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree).
+This rule applies to any element with the [semantic role](#semantic-role) of heading that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree), and that either contains text nodes that do not only consist of [Unicode separator characters](https://www.unicode.org/versions/Unicode11.0.0/ch04.pdf#G134153), or has an accessible name that does not only consist of Unicode separator characters.
```
Comment (jkodu) [inline _rules/SC2-4-6-descriptive-headings.md]: Suggestion:

...and does not only consist of [Unicode separator characters]...

**Comment by annethyme** [comment]
I have misplaced my test asset, "oranges.png". I don't know how to get it into a folder... Can @jkodu help?

**Comment by annethyme** [comment]
@ShadowBB,
As I see it, this success criterion is for sighted users as well as screen reader users, so both user groups deserve a descriptive heading.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 2]
```diff
@@ -22,18 +23,26 @@ authors:
...
 
-**Note**: This rule only applies to elements with the [semantic role](#semantic-role) of heading. Thus, it is a partial check for WCAG 2.0 success criterion 2.4.6, which applies to all headings. "Heading" is used in its general sense and includes headlines and other ways to add a heading to different types of content. This includes elements that are not marked up as headings in the code, but still act visually as headings, e.g. by larger and/or bolder text. 
+**Note**: This rule only applies to elements with the [semantic role](#semantic-role) of `heading`. Thus, it is a partial check for WCAG 2.0 success criterion 2.4.6, which applies to all headings. "Heading" is used in its general sense and includes headlines and other ways to add a heading to different types of content. This includes elements that are not marked up as headings in the code, but still act visually as headings, e.g. by larger and/or bolder text.
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: You could perhaps shorten this to...
This rule only applies to elements with the [semantic role](#semantic-role) of `heading`. Thus, it is a partial check for WCAG 2.0 success criterion 2.4.6, which applies to all headings regardless of semantic programming. In the success criteria "heading" is used in the general sense and includes elements that are not marked up as headings in the code, but still act visually as headings, e.g. by larger and/or bolder text.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 3]
```diff
@@ -22,18 +23,26 @@ authors:
...
+**Note**: This rule only applies to elements with the [semantic role](#semantic-role) of `heading`. Thus, it is a partial check for WCAG 2.0 success criterion 2.4.6, which applies to all headings. "Heading" is used in its general sense and includes headlines and other ways to add a heading to different types of content. This includes elements that are not marked up as headings in the code, but still act visually as headings, e.g. by larger and/or bolder text. 
 
-### Expectation
+### Expectation 1
 
-Each target element describes the topic or purpose of its [section of the content](#section-of-content).
+The [visible](#visible) [content](https://www.w3.org/TR/WCAG21/#dfn-content) of each target element describes the topic or purpose of the entirety or a part of its [section of the content](#section-of-content), or the `heading` has no section of the content.
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: The [visible](#visible) [content](https://www.w3.org/TR/WCAG21/#dfn-content) of each target element describes the topic or purpose of the entirety or a part of its [section of the content](#section-of-content), or the `heading` has no section of the content.

change to...

The [visible](#visible) [content](https://www.w3.org/TR/WCAG21/#dfn-content) of each target element describes the topic or purpose of its [section of the content](#section-of-content) in part or in it's entirety, or the `heading` has no section of the content.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 4]
```diff
@@ -22,18 +23,26 @@ authors:
...
 
 **Note**: Headings do not need to be lengthy. A word, or even a single character, may suffice.
 
+**Note:** Having a heading that has no section of content might be a WCAG violation under other success criteria (e.g. 1.3.1 Info and Relationships, if the heading markup has been misused for purely presentational purposes), however, this is out of scope for this rule. 
+
+### Expectation 2
+
+The [accessible name](#accessible-name) of each target element describes the topic or purpose of the entirety or a part of its [section of the content](#section-of-content), or the `heading` has no section of the content.
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: See previous comment, this doesn't sound right... "...describes the topic or purpose of the entirety or a part of its [section of the content](#section-of-content)."

File: _rules/SC2-4-6-descriptive-headings.md [hunk 5]
```diff
@@ -22,18 +23,26 @@ authors:
...
+
+**Note:** Usually the accessible name will be the same as the visible content of the `heading`, but in some cases the two might be different, and in these cases both versions of the `heading` need to be evaluated.
+
 ## Assumptions
-_There are currently no assumptions._
+This rule assumes that while having a heading that has no section of content might be a WCAG violation under other success criteria (e.g. 1.3.1 Info and Relationships, if the heading markup has been misused for purely presentational purposes), this is allowed under success criterion 2.4.6 Headings and Labels: Headings and labels describe topic or purpose.
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: Is this duplication of what has already been said under Expectation 1?

File: _rules/SC2-4-6-descriptive-headings.md [hunk 6]
```diff
@@ -110,6 +119,42 @@ Heading marked up with `h` element that describes the topic or purpose of its se
 <p>We are open Monday through Friday from 10 to 16</p>
 ```
 
+#### Passed example 7
+
+Heading marked up with `h` element that describes the topic or purpose of a part of its section of the content.
+
+```html
+<h1 class="target">Oranges</h1>
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: What is the target class for? It seems redundant.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 7]
```diff
@@ -110,6 +119,42 @@ Heading marked up with `h` element that describes the topic or purpose of its se
...
+Heading marked up with `h` element where both the visible text nodes as well as the accessible name describes the content, though the two are different.
+
+```html
+<h1 class="target" aria-label="Office opening hours">Opening Hours</h1>
+<p>We are open Monday through Friday from 10 to 16</p>
+```
+
+#### Passed example 10
+
+Visible heading is made up by an image of text, that also has an accessible name. Both visible content and accessible name is descriptive.
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: **The** visible heading is made up **of** an image of text, that also has an accessible name. Both **the** visible content and **the** accessible name describe the content.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 8]
```diff
@@ -150,18 +195,21 @@ Heading marked up with `h` element that does not describe the topic or purpose o
 
 #### Failed example 5
 
-Empty heading marked up with `h` element.
+Heading marked up with `h` element where the visible text nodes describe the content, but the accessible name doesn't.
```
Comment (Brynanders) [inline _rules/SC2-4-6-descriptive-headings.md]: Not sure if I should be flagging this stuff but it seems like you are using a kind of short form of English
Heading marked up with `h` element...
as opposed to 
A heading is marked up with an `h` element...

File: _rules/SC2-4-6-descriptive-headings.md (hunk 2 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: @Brynanders, I expect this note to be redundant as soon as we update to the newest version of the ACT Rules Format, that has a section "outcome mapping" under the "Accessibility requirements" section.
For now, I have left the note as suggested by Difi.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 5 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: @Brynanders: YES, I apparently forgot to delete it!

File: _rules/SC2-4-6-descriptive-headings.md (hunk 6 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: No clue, it comes from the test case I copied+modified. Many of the test cases have `class="target"` on them. I will remove it.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 8 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: It was there in the rule when I started editing it. I'll change it if the native English speaker says so :)

File: _rules/SC2-4-6-descriptive-headings.md (hunk 5 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: deleting the note, it is...

File: _rules/SC2-4-6-descriptive-headings.md [hunk 9]
```diff
@@ -22,18 +23,24 @@ authors:
...
+### Expectation 1
 
-Each target element describes the topic or purpose of its [section of the content](#section-of-content).
+The [visible](#visible) [content](https://www.w3.org/TR/WCAG21/#dfn-content), if there is any, of each target element describes the topic or purpose of its [section of the content](#section-of-content) in part or in its entirety, or the `heading` has no section of the content.
 
 **Note**: Headings do not need to be lengthy. A word, or even a single character, may suffice.
 
+### Expectation 2
+
+The [accessible name](#accessible-name), if there is any, of each target element describes the topic or purpose of its [section of the content](#section-of-content) in part or its entirety, or the `heading` has no section of the content.
```
Comment (audreymaniez) [inline _rules/SC2-4-6-descriptive-headings.md]: I understand the "if there is any" clause for the "visible content", but not for the accessible name.  How a heading could not have an accessible name ?

File: _rules/SC2-4-6-descriptive-headings.md [hunk 10]
```diff
@@ -176,9 +222,25 @@ No heading.
...
+```html
+<h1></h1>
+```
+
+#### Inapplicable example 4
+
+The heading marked up with `role="heading"` does not contain visible content or has an accessible name that does not only consist of [Unicode separator characters](https://www.unicode.org/versions/Unicode11.0.0/ch04.pdf#G134153).
+
+```html
+<h1 aria-label=" "></h1>
```
Comment (DagfinnRomen) [inline _rules/SC2-4-6-descriptive-headings.md]: In the markup, the heading does not role="heading". I suggest you use the span element in stead of the h element to mirror the other test cases, such as failed example 3.

Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: Well spotted! Corrected now...

File: _rules/SC2-4-6-descriptive-headings.md (hunk 2 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: Note changed to @Brynanders's suggestion.

**Comment by annethyme** [comment]
I have sent out an email with a "Final Call" for this rule.
The final call is scheduled to end on April 17, unless extension is requested.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 11]
```diff
@@ -103,27 +110,63 @@ Heading marked up with `role="heading"` that describes the topic or purpose of i
...
+<p>Apples are great too, though.</p>
+```
+
+#### Passed example 8
+
+The heading marked up with `h` element doesn't have a section of content.
+
+```html
+<h1>Oranges on sale</h1>
+```
```
Comment (EmmaJP) [inline _rules/SC2-4-6-descriptive-headings.md]: For me this would fail our check that all headings are followed by some content, so it feels uncomfortable to have it passing a test to say it describes the topic or purpose of something that isn't there.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 12]
```diff
@@ -103,27 +110,63 @@ Heading marked up with `role="heading"` that describes the topic or purpose of i
...
 
 ```html
-<h1 class="target" aria-hidden="true">Opening Hours</h1>
+<h1 aria-label="Office opening hours">Opening Hours</h1>
 <p>We are open Monday through Friday from 10 to 16</p>
 ```
 
+#### Passed example 10
+
+The visible heading is made up of an image of text, that also has an accessible name. Both the visible content and the accessible name is describe the content.
```
Comment (EmmaJP) [inline _rules/SC2-4-6-descriptive-headings.md]: Grammar ... "Both the visible content and the accessible name describe the content." 

Remove "is"

File: _rules/SC2-4-6-descriptive-headings.md [hunk 13]
```diff
@@ -141,27 +184,30 @@ Heading marked up with `role="heading"` that does not describe the topic or purp
...
 ```html
-<h1 class="target" aria-hidden="true">Weather</h1>
+<h1 aria-hidden="true">Weather</h1>
 <p>We are open Monday through Friday from 10 to 16</p>
 ```
 
 #### Failed example 5
 
-Empty heading marked up with `h` element.
+The heading marked up with `h` element has visible text nodes that describe the content, but the accessible name doesn't.
```
Comment (EmmaJP) [inline _rules/SC2-4-6-descriptive-headings.md]: Grammar ... should this read "visible text nodes" or "a visible text node"?

File: _rules/SC2-4-6-descriptive-headings.md [hunk 14]
```diff
@@ -141,27 +184,30 @@ Heading marked up with `role="heading"` that does not describe the topic or purp
...
-<h1></h1>
+<h1 aria-labelledby="id1">Opening hours</h1>
+<p>We are open Monday through Friday from 10 to 16</p>
+<p id="id1">Weather</p>
 ```
 
 #### Failed example 6
 
-Empty heading marked up with `role="heading"`.
+The heading marked up with `h` element has an accessible name that describes the content, but the visible text nodes don't.
```
Comment (EmmaJP) [inline _rules/SC2-4-6-descriptive-headings.md]: Grammar ... should this read "visible text nodes don't" or "visible text node doesn't"?

File: _rules/SC2-4-6-descriptive-headings.md [hunk 15]
```diff
@@ -94,7 +101,7 @@ Heading marked up with `h` element that is a single character that describes the
 
 #### Passed example 5
 
-Heading marked up with `role="heading"` that describes the topic or purpose of its section of the content. The heading is positioned off screen and is included in the Accessibility Tree.
+The heading marked up with `role="heading"` describes the topic or purpose of its section of the content. The heading is positioned off screen and is included in the Accessibility Tree.
```
Comment (audreymaniez) [inline _rules/SC2-4-6-descriptive-headings.md]: What if the heading is visible and included in the accessibility tree and the section is positionned off screen ?

````
<span role="heading">Opening Hours</span>
<p style="position: absolute; top: -9999px; left: -9999px;">We are open Monday through Friday from 10 to 16</p>
````

Would you consider it a passed example ? Or does it fail into the case of "no section of the content" since the content is not visible ?

Is the expectation 2 valid in that configuration ? 
 
> The visible content, if there is any, of each target element describes the topic or purpose of its section of the content in part or in its entirety, or the heading has no section of the content.

Don't we need to be a little more precise on the visibility/state of the "section of the content", as you've done perfectly for the heading ? 

Does the "section of the content" has to be "visible" ? Or does a "section of the content" only need to be in the accessibility tree ? 
and... what if the section of the content is visible but not included in the accessibility tree ?

File: _rules/SC2-4-6-descriptive-headings.md [hunk 16]
```diff
@@ -22,18 +23,24 @@ authors:
 
 ### Applicability
 
-This rule applies to any element with the [semantic role](#semantic-role) of heading that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree).
+This rule applies to any element with the [semantic role](#semantic-role) of `heading` that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree), and that either has [visible](#visible) [content](https://www.w3.org/TR/WCAG21/#dfn-content), or has an [accessible name](#accessible-name) that does not only consist of [Unicode separator characters](https://www.unicode.org/versions/Unicode11.0.0/ch04.pdf#G134153).
```
Comment (WilcoFiers) [inline _rules/SC2-4-6-descriptive-headings.md]: Sorry, but the definition of "content" is subjective and so can't be used in applicability. I don't think we can create an objective definition of what WCAG considers "content". I'm not sure I have a good solution for this. The best thing I've come up with in thinking about this for 15 min is to just look at any visible heading, or any heading with an accessible name. A heading might be visible without it having content, for instance if it was just a line or something, but the expectations would still be valid.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 17]
```diff
@@ -71,19 +78,19 @@ Heading marked up with `role="heading"` that describes the topic or purpose of i
 
 #### Passed example 3
 
-Heading marked up with `h` element with an image that describes the topic or purpose of its section of the content.
+The heading marked up with `h` element contains an image and an accessible name that describes the topic or purpose of its section of the content.
 
 ```html
-<h1 class="target"><img scr="../test-assets/opening_hours_icon.png" alt="Opening hours"></img></h1>
+<h1><img scr="../test-assets/opening_hours_icon.png" alt="Opening hours"></img></h1>
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply _rules/SC2-4-6-descriptive-headings.md]: Yes it should be. This is invalid HTML.

**Comment by ShadowBB** [comment]
> 
> 
> I think this rule should be split up. You have clearly two "types" of headings you want to test, one with visible content, and one with an accessible name, and you have an expectation for each of those.

Isn't example 9 a heading that is both (has both visible conent and an accessible name)? In theory the 2 expectations can simply be merged and state the visible content and the accessible name describe the topic or purpose, correct?

File: _rules/SC2-4-6-descriptive-headings.md (hunk 16 above)
Comment (ShadowBB) [reply _rules/SC2-4-6-descriptive-headings.md]: I think this suggestion would work. Simply look at all visible headings.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 11 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: @ShadowBB, this edit came from you guys. Can you chime in with any arguments on why you feel a heading without content underneath it is not an issue?

**Comment by annethyme** [comment]
I tried splitting it out in two rules where one will only test things that haven't been tested in the other rule, but regretted splitting it up like this after looking into Audrey's comment about defining whether the content is also either visible or included in the accessibility tree. 

Here are the split up (but soon to be reverted) rules for references: 

## Rule no. 1

> name: Heading is descriptive
> description: |
>    This rule checks that visible headings describe the topic or purpose of the content.
> 
> ### Applicability
> 
> This rule applies to any [visible](#visible) element with the [semantic role](#semantic-role) of `heading`.
> 
> **Note:**** Headings with only [whitespace](#whitespace) are not [visible](#visible).
> 
> ### Expectation 1
> 
> The [visible](#visible) heading of each target element describes the topic or purpose of its [section of the content](#section-of-content) in part or in its entirety, or the `heading` has no section of the content.

## Rule no. 2

> name: Heading accessible name is descriptive
> 
> description: |
>    This rule checks that the accessible name for headings describe the topic or purpose of the content.
> 
> ## Test procedure
> 
> ### Applicability
> 
> This rule applies to any HTML or SVG element with the [semantic role](#semantic-role) of `heading` that:
> - is [included in the accessibility tree](#included-in-the-accessibility-tree),
> - has an [accessible name](#accessible-name) that is not only [whitespace](#whitespace), and
> - does not have any [visible](#visible) [text nodes](), or where the [visible](#visible) [text nodes]() does not [match](#matching-characters) the [accessible name](#accessible-name).
> 
> ### Expectation 1
> 
> The [accessible name](#accessible-name) of each target element describes the topic or purpose of its [section of the content](#section-of-content) in part or its entirety, or the `heading` has no section of the content.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 13 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: There can apparently be more text nodes, so we are going with the plural form.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 14 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: There can apparently be more text nodes, so we are going with the plural form.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 18]
```diff
@@ -103,27 +110,63 @@ Heading marked up with `role="heading"` that describes the topic or purpose of i
 
 #### Passed example 6
 
-Heading marked up with `h` element that describes the topic or purpose of its section of the content. The heading is visible, but is not included in the Accessibility Tree.
+The heading marked up with `h` element describes the topic or purpose of its section of the content. The heading is visible, but is not included in the Accessibility Tree.
+
+```html
+<h1 aria-hidden="true">Opening Hours</h1>
+<p>We are open Monday through Friday from 10 to 16</p>
+```
```
[the remark this answers is not in the document]
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: @EmmaJP, I am not trying to illustrate best practice here, just trying to create extensive test case coverage for how the rule is supposed to work. 
If anyone has good ideas for how to improve the test case description, I am more than willing to add it.
But I don't personally know that there are any good use cases for this, so I don't know what to try to explain.
However, judging whether the heading should have been included in the accessibility tree or not is out of scope for this rule, so unless we can quickly throw something together, I'd rather not spend too much time on trying to come up with something for it.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 15 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: @audreymaniez, I think you are perfectly right about specifying if the content has to be visible or included in the accessibility tree. 
However, the "no section of content" makes everything pass anyway, so I think we need to have another look at that part too. I haven't updated the test cases yet, since a decision on the "no section content" part can change the test cases quite a lot.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 11 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: @EmmaJP, I guess the argument is actually that if the heading markup has been misused to mark something up as a heading though it actually isn't, this would be a failure under succes criterion 1.3.1 Info and Relationships, but not per se under 2.4.6 Headings and Labels.

@kasperisager and I just discussed if there should actually be 2 expectations (per current expectation) that would roughly be something like this:

1. The heading has a section of content
2. The heading is descriptive of its section of content  

... but after thinking about it, I think the first one is actually a 1.3.1 violation.

I suggest keeping the "or doesn't have a section of content" part.

**Comment by annethyme** [comment]
@WilcoFiers, my concern about splitting the rule up in two is to have two semi-automated rules that potentially overlap in all cases where the accessible name is taken directly from the visible text nodes of the heading. 
Collecting it in one rule makes it easier to quickly decide if there is a difference, and each one needs to be reviewed separately.
I tried yesterday splitting it out in two rules that *did not overlap*, but Audrey's comment about the section of content being visible/included in the accessibility tree rules out this approach.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 11 above)
Comment (EmmaJP) [reply _rules/SC2-4-6-descriptive-headings.md]: Surely then this should be 'not applicable', as there is no content to describe, rather than a pass.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (EmmaJP) [reply _rules/SC2-4-6-descriptive-headings.md]: About the only situation I can think of would be when there is a visible heading and a non-visible heading for the same content, that are slightly different.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 11 above)
Comment (annethyme) [reply _rules/SC2-4-6-descriptive-headings.md]: Right now the "or doesn't have a section of content" part lives in the expectation, so as the rule is written now, it is a Pass.

If we move this to the Applicability, you will already in the applicability have to judge whether the heading is connected to a section of content or not. 
At least that will limit the possibilities for automating the applicability.

**Comment by annethyme** [comment]
AGENDA ITEM fo CG call on April 25: 

Is it better to split this rule up in two or keep it as one, as per this discussion (and onwards): 
https://github.com/act-rules/act-rules.github.io/pull/444#pullrequestreview-222704240

File: _rules/SC2-4-6-descriptive-headings.md [hunk 19]
```diff
@@ -47,22 +58,22 @@ _There are no major accessibility support issues known for this rule._
 - [ARIA12: Using role=heading to identify headings](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA12)
 - [HTML 5.2 Standard - Heading content](https://www.w3.org/TR/html52/dom.html#heading-content)
 
-## Test Cases
+## Test Cases - UPDATE IN PROGRESS
```
Comment (annethyme) [inline _rules/SC2-4-6-descriptive-headings.md]: @annethyme, make sure to update this.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 11 above)
Comment (ShadowBB) [reply _rules/SC2-4-6-descriptive-headings.md]: Sorry for chiming in to this so late.

The rule should not fail this situation because the rule doesn't fail 1.3.1 and this situation should fail 1.3.1.

This situation should not be not applicable because it is not possible to create an applicability that differentiates between headings with content and headings without content. This can be done in the expectation part of the rule, but can't be done in the applicability part of the rule.

**Comment by annethyme** [comment]
Taking the rule out of Final Call since substantial changes are expected.

File: _rules/SC2-4-6-descriptive-headings.md (hunk 11 above)
Comment (EmmaJP) [reply _rules/SC2-4-6-descriptive-headings.md]: Sorry to come back to this after so long. This one is tricky. 

I guess the heading could be of a kind where no content following it would be meaningful. That is not necessarily true of just any heading text.

Something like "Delivery availability" could work with no content below. Perhaps "Oranges on sale" is such a heading, but it isn't obviously so.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 20]
```diff
@@ -92,93 +106,195 @@ Heading marked up with `h` element that is a single character that describes the
...
+heading is visible, but not included in accessibility tree
+content is not visible, but included in the a accessibility tree
+heading is not descriptive, but it doesn't matter
+
+#### Passed Example 13
+
+heading is included in accessibility tree, but not visible
+content is visible, but not included in accessibility tree
+heading is not descriptive, but it doesn't matter
+
```
Comment (EmmaJP) [inline _rules/SC2-4-6-descriptive-headings.md]: That is a very odd situation!

**Comment by annethyme** [comment]
Suggestion for splitting up this rule in two:

# Suggested rule 1: Visible heading is descriptive

## Applicability

This rule applies to any element with the [semantic role](#semantic-role) of `heading` that is  [visible](#visible).

## Expectation

The [visible](#visible) heading of each target element describes the topic or purpose of its [visible](#visible) [section of the content](#section-of-content) in part or in its entirety, or the `heading` has no [visible](#visible) [section of the content](#section-of-content).

# Suggested rule 2: Accessible name of heading is descriptive

## Applicability

This rule applies to any element with the [semantic role](#semantic-role) of `heading` that is [included in the accessibility tree](#included-in-the-accessibility-tree), has an [accessible name](#accessible-name) that is not only [whitespace](#whitespace), and where the:
- [accessible name](#accessible-name) is different than the [visible](#visible) heading, OR 
- where its [section of the content](#section-of-content) that is [included in the accessibility tree](#included-in-the-accessibility-tree) is different from its [visible](#visible) [section of the content](#section-of-content).

## Expectation

The [accessible name](#accessible-name) of each target element describes the topic or purpose of its [section of the content](#section-of-content) that is [included in the accessibility tree](#included-in-the-accessibility-tree) in part or its entirety, or the `heading` has no [section of the content](#section-of-content) that is [included in the accessibility tree](#included-in-the-accessibility-tree).

**Comment by annethyme** [comment]
@kasperisager is on board with the suggested split up, so @annethyme will split up the rule to get it ready for review soon again.

File: _rules/heading-accessible-name-is-descriptive-fdaadc.md [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+
+## Background
+
+- [Understanding Success Criterion 2.4.6: Headings and Labels](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html) 
+- [G130: Providing descriptive headings](https://www.w3.org/WAI/WCAG21/Techniques/general/G130)
+- [H42: Using h1-h6 to identify headings](https://www.w3.org/WAI/WCAG21/Techniques/html/H42)
+- [ARIA12: Using role=heading to identify headings](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA12)
+- [HTML 5.2 Standard - Heading content](https://www.w3.org/TR/html52/dom.html#heading-content)
+
+## Test Cases - UPDATE IN PROGRESS
```
Comment (annethyme) [inline _rules/heading-accessible-name-is-descriptive-fdaadc.md]: @annethyme, make sure to update this

File: _rules/heading-accessible-name-is-descriptive-fdaadc.md [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+- Anne Thyme Nørregaard
+- Dagfinn Rømen
+- Geir Sindre Fossøy
+- Carlos Duarte
+---
+
+## Applicability
+
+This rule applies to any element with the [semantic role](#semantic-role) of `heading` that is [included in the accessibility tree](#included-in-the-accessibility-tree), has an [accessible name](#accessible-name) that is not only [whitespace](#whitespace), and where:
+- the [accessible name](#accessible-name) is different than the [visible](#visible) heading, OR
```
Comment (annethyme) [inline _rules/heading-accessible-name-is-descriptive-fdaadc.md]: - the [accessible name](#accessible-name) does not [match](#matching-characters) the combined [characters](#character) formed by the [visible](#visible) [descendant text content](#) and [characters](#character) in [images of text](https://www.w3.org/TR/WCAG21/#dfn-images-of-text).

File: _rules/heading-accessible-name-is-descriptive-fdaadc.md [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+- Anne Thyme Nørregaard
+- Dagfinn Rømen
+- Geir Sindre Fossøy
+- Carlos Duarte
+---
+
+## Applicability
+
+This rule applies to any element with the [semantic role](#semantic-role) of `heading` that is [included in the accessibility tree](#included-in-the-accessibility-tree), has an [accessible name](#accessible-name) that is not only [whitespace](#whitespace), and where:
+- the [accessible name](#accessible-name) does not [match](#matching-characters) the [visible](#visible) heading that is the combined [characters](#character) formed by the [visible](#visible) [descendant text content](#) and [characters](#character) in [images of text](https://www.w3.org/TR/WCAG21/#dfn-images-of-text), OR
```
Comment (annethyme) [inline _rules/heading-accessible-name-is-descriptive-fdaadc.md]: [visible](#visible) [descendant text content](#) --> [visible](#visible) [shadow-including] [descendant text content](#) - except that the descendant text content definition doesn't work in this case

images of text also needs to be visible shadow-including descendants

Comment (annethyme) [reply _rules/heading-accessible-name-is-descriptive-fdaadc.md]: flat tree text content: the concatenation of the data of all text node descendants of node in the flat tree, in tree order.

File: _rules/SC2-4-6-descriptive-headings.md [hunk 24]
```diff
@@ -1,39 +1,45 @@
...
 ---
 
-## Test procedure
+## Applicability
 
-### Applicability
+This rule applies to any [visible](#visible) element with the [semantic role](#semantic-role) of `heading`.
 
-This rule applies to any element with the [semantic role](#semantic-role) of heading that is either [visible](#visible) or [included in the accessibility tree](#included-in-the-accessibility-tree).
+**Note:**** Headings with only [whitespace](#whitespace) are not [visible](#visible).
```
Comment (annethyme) [inline _rules/SC2-4-6-descriptive-headings.md]: remove `**`

File: _rules/SC2-4-6-descriptive-headings.md [hunk 25]
```diff
@@ -1,39 +1,45 @@
...
 
-**Note**: This rule only applies to elements with the [semantic role](#semantic-role) of heading. Thus, it is a partial check for WCAG 2.0 success criterion 2.4.6, which applies to all headings. "Heading" is used in its general sense and includes headlines and other ways to add a heading to different types of content. This includes elements that are not marked up as headings in the code, but still act visually as headings, e.g. by larger and/or bolder text. 
+## Expectation
 
-### Expectation
+The [visible](#visible) heading of each target element describes the topic or purpose of its [visible](#visible) [section of the content](#section-of-content) in part or in its entirety, or the `heading` has no [visible](#visible) [section of the content](#section-of-content).
 
-Each target element describes the topic or purpose of its [section of the content](#section-of-content).
+The [visible](#visible) heading is the combined [characters](#character) formed by the [visible](#visible) [descendant text content](#) and [characters](#character) in [images of text](https://www.w3.org/TR/WCAG21/#dfn-images-of-text).
```
Comment (annethyme) [inline _rules/SC2-4-6-descriptive-headings.md]: Remove link to character and insert note about definition missing + create issue on character definition.

File: _rules/heading-accessible-name-is-descriptive-fdaadc.md [hunk 26]
```diff
@@ -0,0 +1,290 @@
+---
```
Comment (annethyme) [inline _rules/heading-accessible-name-is-descriptive-fdaadc.md]: Split this rule out into separate PR.

File: pages/glossary/character.md [hunk 27]
```diff
@@ -0,0 +1,29 @@
+---
```
Comment (annethyme) [inline pages/glossary/character.md]: Remove definition from this PR

**Comment by WilcoFiers** [comment]
Closing due to inactivity.
